p2p/host/basic/basic_host_test.go

```diff
@@ -91,16 +96,35 @@ func TestProtocolHandlerEvents(t *testing.T) {
 	}
 	defer sub.Close()
 
-	assert := func(added, removed []protocol.ID) {
-		var next event.EvtLocalProtocolsUpdated
-		select {
-		case evt := <-sub.Out():
-			next = evt.(event.EvtLocalProtocolsUpdated)
-			break
-		case <-time.After(5 * time.Second):
-			t.Fatal("event not received in 5 seconds")
+	// the identify service adds new protocol handlers shortly after the host
+	// starts. this helps us filter those events out, since they're unrelated
+	// to the test.
+	isIdentify := func(evt event.EvtLocalProtocolsUpdated) bool {
+		for _, p := range evt.Added {
+			if p == identify.ID || p == identify.IDPush {
+				return true
+			}
+		}
+		return false
+	}
+
+	nextEvent := func() event.EvtLocalProtocolsUpdated {
+		for {
+			select {
+			case evt := <-sub.Out():
+				next := evt.(event.EvtLocalProtocolsUpdated)
+				if isIdentify(next) {
+					continue
+				}
+				return next
+			case <-time.After(5 * time.Second):
+				t.Fatal("event not received in 5 seconds")
```

Comment:
Fatal must be called from the main goroutine: https://golang.org/pkg/testing/#T

Reply:
This code is being executed on the main goroutine, not on a separate one.